micro-rdk/src/common/provisioning/server.rs

```diff
@@ -0,0 +1,528 @@
+where
+    S: Storage + Clone + 'static,
+    GrpcError: From<S::Error>,
+{
+    _exec: PhantomData<E>,
+    _stream: PhantomData<I>,
+    _storage: PhantomData<S>,
+    #[pin]
+    connection: http2::Connection<I, ProvisioningService<S>, E>,
+    credential_ready: AtomicSync,
+}
+
+impl<I, S, E> Future for ProvisoningServer<I, S, E>
+where
+    S: Storage + Clone + 'static,
+    I: rt::Read + rt::Write + std::marker::Unpin + 'static,
+    GrpcError: From<S::Error>,
+    E: rt::bounds::Http2ServerConnExec<
+        <ProvisioningService<S> as Service<Request<Incoming>>>::Future,
+        GrpcBody,
+    >,
+{
+    type Output = Result<(), hyper::Error>;
+    fn poll(
+        self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        let this = self.project();
+
+        match this.connection.poll(cx) {
+            Poll::Pending => futures_lite::ready!(this.credential_ready.poll(cx)),
+            Poll::Ready(c) => return Poll::Ready(c),
+        }
+        Poll::Ready(Ok(()))
+    }
+}
+
+impl<I, S, E> ProvisoningServer<I, S, E>
+where
+    S: Storage + Clone + 'static,
+    I: rt::Read + rt::Write + std::marker::Unpin + 'static,
+    GrpcError: From<S::Error>,
+    E: rt::bounds::Http2ServerConnExec<
+        <ProvisioningService<S> as Service<Request<Incoming>>>::Future,
+        GrpcBody,
+    >,
+{
+    fn new(service: ProvisioningService<S>, executor: E, stream: I) -> Self {
+        let credential_ready = service.get_credential_ready();
+        let connection = http2::Builder::new(executor).serve_connection(stream, service);
+        Self {
+            _exec: PhantomData::default(),
+            _stream: PhantomData::default(),
+            _storage: PhantomData::default(),
+            connection,
+            credential_ready,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{
+        net::{Ipv4Addr, SocketAddr, TcpListener, TcpStream},
+        time::Duration,
+    };
+
+    use async_io::{Async, Timer};
+
+    use http_body_util::BodyExt;
+    use http_body_util::Full;
+    use hyper::{
+        header::{CONTENT_TYPE, TE},
+        Method,
+    };
+    use prost::Message;
+
+    use crate::{
+        common::{
+            app_client::encode_request,
+            provisioning::{
+                server::{ProvisioningInfo, ProvisioningServiceBuilder, ProvisoningServer},
+                storage::{MemoryCredentialStorage, Storage},
+            },
+        },
+        native::{exec::NativeExecutor, tcp::NativeStream},
+        proto::provisioning::v1::{
+            CloudConfig, GetNetworkListRequest, GetSmartMachineStatusRequest,
+            GetSmartMachineStatusResponse, SetSmartMachineCredentialsRequest,
+        },
+    };
+
+    use super::ProvisioningService;
+
+    async fn run_provisioning_server(
+        ex: NativeExecutor,
+        srv: ProvisioningService<MemoryCredentialStorage>,
+    ) {
+        let listen = TcpListener::bind("127.0.0.1:56432");
```

Comment:
maybe make this url & port a constant since it has repeated use?

Reply:
+1